roles/srv-announcer/vars/main.yml

```diff
@@ -0,0 +1,17 @@
+# NOTE: variables set here are considered to be CONSTANTS due to high precedence
```

Comment:
I don't understand the constants/not constants distinction TBH.

Given `announcer_artifact_basename` contains the version, does this mean one can't ever update the version?

Reply:
The idea here is that while [`announcer_version`](https://github.com/wireapp/ansible-sft/pull/8/files#diff-00ee6cc36478d9ce7f01fb88808ddb9fR12) can (and should) be defined by the user of this role, `announcer_artifact_basename`, on the other hand, is defined by the role author or even the entity in charge of defining the [name of the artifact](https://github.com/wireapp/srv-announcer/blob/master/.github/workflows/release-and-upload.yml#L37).

Distinguishing based on the use cases led to putting certain vars either in `./defaults` or `./vars` to make the ones that should not be touched by the user de facto read-only.

Does this makes it a bit more clear?

Reply:
Can we please move everything underneath vars/main.yml to defaults/main.yml?

We don't gain anything from hiding these variables and making them impossible to override. Clearly, for version v0.1.1 of the announcer, we want to override the shasum. And this is a very valid usecase as a `user` of the role, rather than as a role maintainer. Blocking your users of this role from using a newer artifact until you as the maintainer got the time to add a shasum and a new release doesn't make sense to me.

this is just one example. It's hard to know in advance where users may have a need to override a variable. Requiring a role update for any change is not a good practice. Let's please not make our lives more complicated than they need to be; and move these variables to defaults/main.yml for flexibility. If a role user 'mistakenly' overrides a variable that you don't want them to override, it's their problem, not yours. So why forbid this override?

Reply:
Apparently I did a bad job in explaining/documenting the behaviour you are questioning.

There is no need of overriding a checksum. The user has [complete flexibility](https://github.com/wireapp/ansible-sft/blob/add-srv-announcer-role/roles/srv-announcer/defaults/main.yml#L6-L16) in determining the source of the artifact. 

Case 1: if the user provides a path to the artifact, it is assumed that she already verified its integrity before invoking the role. After all, when providing a path the artifact has already been downloaded

Case 2: the user only provides a version, therefore the role needs to be aware of the corresponding checksum, hence [this](https://github.com/wireapp/ansible-sft/blob/add-srv-announcer-role/roles/srv-announcer/vars/main.yml#L17-L19) ever growing map of version --> checksum

Case 3: the user wants to handover the download part to the role, therefore a URL and a checksum are needed. A slightly different scenario - version + checksum - could easily be done outside of the role by deriving the URL from interpolating the version into a URL template string. 

Overall, those cases should cover every scenario. Or would you have one in mind that isn't? 

If you argue against this approach, of using constants, you could question private variables of a class or public vs. private APIs as well - same pattern in that sense.

I can see your point regarding flexibility, although with some reasonable design, I think this won't be the case much often. I actually belief that having both makes it more clear to the reader/user to a) discover the parameters she is actually looking for and wants to provide, and the ones that are internal and would only be noise when looking for these kind of things.

To conclude, if I'm still not able to make a valid case here, I'd be happy to put everything together. Let me know.

Reply:
> Case 2: the user only provides a version, therefore the role needs to be aware of the corresponding checksum, hence this ever growing map of version --> checksum

Well... this is inside vars, so not overridable. If a new release happens in in github for srv announcer, then a user of this role is unable to use option 2 with a newer version of srv announcer. This is a problematic restriction (it's a restriction, it has downsides) and I don't see the benefit of this "constants" approach here. Your comment doesn't explain *why* it's useful to disallow ansible role users from overriding something. What is the benefit? You say - if I get this right - in the scenario where a newer version of an artifact exists and the role isn't aware yet, if option 2 cannot be used, well then the user is free to use option 3 instead. That's nice, but why make things more complicated than they need to be? 

But what about if I want to always have the `latest` version available, whatever that might be?

Other examples:

* srv_announcer is hosted on gitlab going forward. This role has implicit assumptions that github is where srv announcer is found (it's hardcoded here).
* If I fork the srv announcer, I can currently not override the srv_announcer_download_base_url. I would need to *also* fork and maintain this ansible-sft project.
* If I fork the srv announcer, but decide to rename it, I cannot override the artifact basename without first forking this repo
* I want to support 32 bit. Using my forked srv announncer repo. But I cannot do this, I first need to fork this, and move srv_announcer_arch to defaults/main.yml, and then I can override it in group_vars for the 32-bit webserver group

But all this is beside the point. I was trying to come up with a reason to *not restrict users* by giving an example. The example is just that - an example. I cannot really foresee what someone might want to do in 2021 with this role. However, I have in the past been quite frustrated with ansible role developers who unnecessarily placed variables inside /vars for no good reason, and I had to fork the project, instead of being able to simply override one variable - which is already there, in variable form (not hardcoded in 3 places), but it's in an unoverridable location. So I would like us to not start a pattern in which it's hard to override things without forking the code.

> If you argue against this approach, of using constants, you could question private variables of a class or public vs. private APIs as well - same pattern in that sense.

Ansible is not really a programming language, it's a limited DSL with a range of conventions and helpers. It's not the same.

I found another person with a view on this that I find sensible, see for example https://codereviewvideos.com/course/ansible-tutorial/video/variable-precedence-where-to-put-your-role-vars

Is this another case of https://github.com/zinfra/backend-issues/issues/1617 ?

Reply:
(I also opened a separate discussion for this here: https://github.com/zinfra/backend-issues/issues/1707 )

Reply:
For Case 2 - only provide a version - I cannot see a way to make it work w/o hard-coding and therefore releasing a new version of the role every time a new SFT version comes out, because of the checksum. Fetching the checksum (e.g. `get_url`) is pointless.

I think, most of your concerns in that regard can be addressed (or at least be minimized to an absolute minimum) by proper automation timing these things (e.g. artifact publishing and checksum adding + version bumping in the role). 

`latest`: If that is a desired use case - and I hope it's not, because it leads to unpredictable behaviour - this could easily be solved with *case 3* + us providing a URL where the version is replaced by *latest* always pointing to the latest release (similar to container images and the *latest* tag). Tho, it's far from being practical, because of the ever-changing checksum. I intention would be to prevent people from doing these things.

For me, it's about providing the user with a maximum amount of convenience while keeping it flexible, hence `srv_announcer_version`.

>> If you argue against this approach, of using constants, you could question private variables of a class or public vs. private APIs as well - same pattern in that sense.

> Ansible is not really a programming language, it's a limited DSL with a range of conventions and helpers. It's not the same.

There are differences, for sure. But, IMHO, there are also similarities and applicable concepts. Let's agree to disagree :)

> To conclude, if I'm still not able to make a valid case here, I'd be happy to put everything together. Let me know.

I'm not going into every example you brought up, but I see your point. Even though the offer still stands, I'd much rather like to discuss your concerns about specific variables defined in `vars/main.yml` - like `srv_announcer_download_base_url` - instead of just moving everything over.